x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/job/RollupIDGenerator.java

```diff
@@ -0,0 +1,175 @@
+            }
+        }
+
+        @Override
+        public void add(String v) {
+            if (check(v)) {
+                byte[] vs = (v).getBytes(StandardCharsets.UTF_8);
+                crc.update(vs, 0, vs.length);
+            }
+        }
+
+        @Override
+        public void addNull() {
+            // Old ID scheme used (int)19 as the null placeholder.
+            // Not great but we're stuck with it :(
+            crc.update(19);
+        }
+
+        @Override
+        public String getID() {
+            setFlag();
+            return String.valueOf(crc.getValue());
+        }
+    }
+
+    /**
+     * The "new" style ID, used in Rollup V2.  A 128 bit Murmur3 hash of
+     * all the keys concatenated together, base64-encoded, then prepended
+     * with the job ID and a `$` delimiter
+     *
+     * Null values are hashed as a (hopefully) unique string `__NULL_PLACEHOLDER__830f1de2__`
+     */
+    public static class Murmur3 extends RollupIDGenerator {
+        private static final long SEED = 19;
+        private static final BytesRef DELIM = new BytesRef("$".getBytes(StandardCharsets.UTF_8));
+        private static final byte[] NULL_PLACEHOLDER = "__NULL_PLACEHOLDER__830f1de2__".getBytes(StandardCharsets.UTF_8);
+        private final BytesRefBuilder id = new BytesRefBuilder();
+        private final String jobId;
+
+        Murmur3(String jobId) {
+            this.jobId = jobId;
+        }
+
+        @Override
+        public void add(Integer v) {
+            if (check(v)) {
+                update(Numbers.intToBytes(v));
+            }
+        }
+
+        @Override
+        public void add(Long v) {
+            if (check(v)) {
+                update(Numbers.longToBytes(v));
+            }
+        }
+
+        @Override
+        public void add(Double v) {
+            if (check(v)) {
+                update(Numbers.doubleToBytes(v));
+            }
+        }
+
+        @Override
+        public void add(String v) {
+            if (check(v)) {
+                update((v).getBytes(StandardCharsets.UTF_8));
+            }
+        }
+
+        @Override
+        public void addNull() {
+            // New ID scheme uses a (hopefully) unique placeholder for null
+            update(NULL_PLACEHOLDER);
+        }
+
+        private void update(byte[] v) {
+            id.append(v, 0, v.length);
+            id.append(DELIM);
+        }
+
+        @Override
+        public String getID() {
+            setFlag();
+            MurmurHash3.Hash128 hasher
+                = MurmurHash3.hash128(id.bytes(), 0, id.length(), SEED, new MurmurHash3.Hash128());
+            byte[] hashedBytes = new byte[16];
+            System.arraycopy(Numbers.longToBytes(hasher.h1), 0, hashedBytes, 0, 8);
+            System.arraycopy(Numbers.longToBytes(hasher.h2), 0, hashedBytes, 8, 8);
+            return jobId + "$" + Base64.getUrlEncoder().withoutPadding().encodeToString(hashedBytes);
```

Comment:
Since we use 128 bits I think it's ok to just update the hash with the jobId instead of adding it as a prefix ?

Reply:
The prefix compression should handle the jobId efficiently so let s keep it this way. This will prevent collision accross jobs so please forget my last comment ;).

Reply:
Sure, that works for me.  Just need to tweak the tests a bit, since I was using the job ID as a shortcut for "old vs new" ID... but we can use length instead :)

Reply:
For posterity we chatted in slack and decided to leave the job name prepended.  Prefix encoding should make the overhead minimal, and it is convenient to have the job ID attached (as well as preventing some other forms of collisions due to same job but different interval, etc)